ophyd/pseudopos.py

```diff
@@ -105,20 +149,184 @@ def _started_moving(self, value):
+    ----------
+    cls : namedtuple
+        The position class to use. This is likely a RealPosition or a
+        PseudoPosition from a PseudoPositioner.
+    args :
+        User-specified positional arguments
+    kwargs : dict
+        User-specified keyword arguments
+
+    Returns
+    -------
+    position_tuple : cls
+        The position tuple
+    kwargs : dict
+        Keyword arguments not related to the position tuple
+
+    Raises
+    ------
+    TypeError
+        On an empty or invalid namedtuple
+    ValueError
+        On a mismatch of parameters
+    '''
+    usage_info = '''Positions can be passed in a number of ways:
+
+    As positional arguments:
+        pseudo.method(px, py, pz, **kwargs)
+    As a sequence or PseudoPosition/RealPosition:
+        pseudo.method((px, py, pz), **kwargs)
+    As kwargs:
+        pseudo.method(px=1, py=2, pz=3, **kwargs)
+
+    '''
+    try:
+        fields = cls._fields
+    except AttributeError:
+        raise TypeError('Invalid position tuple')
+
+    if not fields:
+        raise TypeError('Invalid position tuple')
+
+    if args and isinstance(args[0], (cls, Sequence)):
+        # Position is in the first positional argument
+        if len(args) > 1:
+            raise ValueError(usage_info +
+                             'Cannot specify more than one positional '
+                             'argument if the first one is a {} '
+                             ''.format(cls.__name__))
+
+        position = args[0]
+        if not isinstance(position, cls):
+            # Ensure a position tuple is passed back
+            position = cls(*position)
+
+        return position, kwargs
+
+    elif len(args) == len(fields):
+        # Position is in positional arguments
+        return cls(*args), kwargs
+
+    elif len(args) > 0:
+        # Position is in positional arguments
+        raise ValueError(usage_info +
+                         'Wrong number of arguments for {}. '
+                         'Got {}, expected {}'
+                         ''.format(cls.__name__, len(args), len(fields)))
+
+    if not kwargs:
+        # no positional arguments or kwargs, just show usage information
+        raise ValueError(usage_info)
+
+    # No positional arguments, position described in terms of kwargs
+    missing_fields = [field for field in fields
+                      if field not in kwargs]
+
+    if missing_fields:
+        raise ValueError(usage_info +
+                         'Missing keyword arguments for field names of {}:'
+                         ' {}'.format(cls.__name__,
+                                      ', '.join(missing_fields)))
+
+    # separate position tuple kwargs from other kwargs
+    position_kw = {field: kwargs[field] for field in fields}
+    other_kw = {key: value for key, value in kwargs.items()
+                if key not in fields}
+    position = cls(**position_kw)
+    return position, other_kw
+
+
+class PseudoPositioner(Device, PositionerBase):
```

Comment:
Can't use the `SoftPositioner` base for `PseudoPositioner` as `SoftPositioner` `move()` adds in things to indicate motion start/stop. `PseudoSingle` works as it never calls `super().move`... Thoughts?